### PR TITLE
[WIP] Add boosted-MMI to chain

### PR DESCRIPTION
**Pull request by hhadian**
Does it look OK? 
If yes, I am going to add the two options boost and hard_boost to chain's train.py and acoustic_model.py.

### CODE REVIEW

File: src/chain/chain-training.h [hunk 1]
```diff
@@ -61,8 +61,18 @@ struct ChainTrainingOptions {
   // should have a softmax as its final nonlinearity.
   BaseFloat xent_regularize;
 
+  // The boosting coefficient (i.e. b in bMMI)
+  BaseFloat boost;
+  
+  // This controls the boosting approach. If true, hard boosting
+  // will be used: the boosting mask consists of 1's and 0's and is 
+  // independent of numerator compuatation (i.e. is a function of numerator
```
Comment (galv) [inline src/chain/chain-training.h]: compuatation -> computation

Comment (galv) [inline src/chain/chain-training.h]: I'm not familiar with boosted MMI, but do you mean "independent of denominator computation"?

File: src/chain/chain-training.h [hunk 2]
```diff
@@ -61,8 +61,18 @@ struct ChainTrainingOptions {
...
+  // graph only).
+  bool hard_boost;
+
   ChainTrainingOptions(): l2_regularize(0.0), leaky_hmm_coefficient(1.0e-05),
-                          xent_regularize(0.0) { }
+                          xent_regularize(0.0), boost(0.1),
```
Comment (galv) [inline src/chain/chain-training.h]: Won't this make boosting on by default, breaking backwards compatibility?

Comment (hhadian) [inline src/chain/chain-training.h]: Oh yes! will fix it.

File: src/chain/chain-training.cc [hunk 3]
```diff
@@ -54,9 +64,24 @@ void ComputeChainObjfAndDeriv(const ChainTrainingOptions &opts,
       numerator.Backward(xent_output_deriv);
     }
   }
+  
+  num_logprob_weighted += -opts.boost * supervision.num_sequences *
+                          supervision.frames_per_sequence;
+  CuMatrixBase<BaseFloat> *boosting_mask = nnet_output_deriv;
+  if (opts.boost != 0.0 && opts.hard_boost) {
+    boosting_mask = new CuMatrix<BaseFloat>(*nnet_output_deriv);
+    boosting_mask->ApplyHeaviside();
```
Comment (danpovey) [inline src/chain/chain-training.cc]: looks fine, but it occurred to me that if the hard mask turns out to be better, we should probably (a) hardcode it to use the hard mask, and (b) we can give the soft version of the mask to the denominator computation and have it do the heaviside itself.  It creates a copy of the matrix anyway, to do the transpose, so we can have it copy the mask before the actual probs, do the heaviside, and then add in the actual probs.  but no need to do this before testing the idea out, the speed difference would be extremely tiny.

Comment (hhadian) [inline src/chain/chain-training.cc]: OK. 
BTW even for the purpose of testing the idea, I think this copying (`new CuMatrix<BaseFloat>(*nnet_output_deriv)`) might fail due to limited memory. I remember I had memory allocation problems with batch-sizes > 128 when I was testing on-GPU numerator computation with **two** nnet_output_deriv matrices.
Anyway I will test it to see what happens.

**Comment by danpovey** [comment]
If you test with a fairly small network this should not be an issue.
But I think the chain training scripts often use a minibatch size of 64.

**Comment by danpovey** [comment]
Since this did not seem to be helpful, I'm closing the pull request.. don't delete the branch, though, in case we want to revisit this in future.
